query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala

```diff
@@ -61,10 +61,15 @@ final case class BinaryJoinExec(queryContext: QueryContext,
 
   protected def args: String = s"binaryOp=$binaryOp, on=$on, ignoring=$ignoring"
 
+  //scalastyle:off method.length
   protected[exec] def compose(childResponses: Observable[(QueryResponse, Int)],
                               firstSchema: Task[ResultSchema],
                               querySession: QuerySession): Observable[RangeVector] = {
     val taskOfResults = childResponses.map {
+      case (QueryResult(_, _, result), _)
+        if (result.size  > queryContext.joinQueryCardLimit && cardinality == Cardinality.OneToOne) =>
+        throw new BadQueryException(s"This query results in more than ${queryContext.joinQueryCardLimit} " +
+          s"join cardinality. Try applying more filters.")
```

Comment:
My concerns with Binary join is not result cardinality - it is in fact the cardinality of the operands. They would cause so much high-card data to be passed over the network.

This check is conducted after the network IO happens, so it may be too late.

Can we push some checks down to the child plans to limit the number of RVs? And can we foresee the number of result vectors early on before scan stage ? 

I am not saying this is easy or possible, we need to evaluate.

Reply:
@vishramachandran yes you are right. i will work on it as a separate PR.